dvc/repo/ls.py

```diff
@@ -0,0 +1,73 @@
+import os
+
+from dvc.exceptions import DvcException, OutputNotFoundError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    from dvc.external_repo import external_repo
+    from dvc.repo import Repo
+    from dvc.utils import relpath
+
+    with external_repo(url, rev) as repo:
+        result = []
+        if isinstance(repo, Repo):
+            result.extend(_ls_outs_repo(repo, target, recursive))
+
+        if not outputs_only:
+            result.extend(_ls_files_repo(repo, target, recursive))
+
+        if target and not len(result):
+            raise DvcException("target {} doesn't exist".format(target))
+
+        def prettify(path_info):
+            if not recursive:
+                prefix = path_info.relpath(repo.root_dir)
+                return prefix.parts[0]
+            return relpath(path_info, repo.root_dir)
+
+        result = list(set(map(prettify, result)))
+        result.sort()
+    return result
+
+
+def _ls_files_repo(repo, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+    from dvc.path_info import PathInfo
+
+    files = []
+    target_path = (
+        os.path.join(repo.root_dir, target) if target else repo.root_dir
+    )
+    tree = CleanTree(WorkingTree(target_path))
+
+    if not os.path.exists(target_path):
+        return []
+
+    if target and not recursive:
+        return [PathInfo(target_path)]
+
+    try:
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            files.extend(map(lambda f: PathInfo(dirpath, f), filenames))
+            if not recursive:
+                files.extend(map(lambda d: PathInfo(dirpath, d), dirnames))
```

Comment:
This is probably the cause of that bug. This line should be outside of `if`, right?